server/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java

```diff
@@ -76,21 +79,24 @@
 
     private int totalShardCount = 0;
 
-    private final Map<String, Set<String>> attributeValuesByAttribute = new HashMap<>();
-    private final Map<String, Recoveries> recoveriesPerNode = new HashMap<>();
+    private final Map<String, Set<String>> attributeValuesByAttribute;
+    private final Map<String, Recoveries> recoveriesPerNode;
 
-    public RoutingNodes(ClusterState clusterState) {
-        this(clusterState, true);
+    public RoutingNodes(RoutingTable routingTable, DiscoveryNodes discoveryNodes) {
+        this(routingTable, discoveryNodes, true);
     }
 
-    public RoutingNodes(ClusterState clusterState, boolean readOnly) {
+    public RoutingNodes(RoutingTable routingTable, DiscoveryNodes discoveryNodes, boolean readOnly) {
```

Comment:
I wonder if we should make these private and add factory methods with names signaling the read-only-ness as well as that they are only to be called by `ClusterState`.

Reply:
Sure moved to two separate factory methods here, looks much cleaner indeed.